File: tests/typesafety/typescript/result_chain.ts
// expect: passes

type Ok<T> = { readonly kind: "ok"; readonly value: T };
type Err = { readonly kind: "err"; readonly reason: string };
type Result<T> = Ok<T> | Err;

function ok<T>(value: T): Result<T> {
  return { kind: "ok", value };
}

function err(reason: string): Result<never> {
  return { kind: "err", reason };
}

function andThen<T, U>(result: Result<T>, step: (value: T) => Result<U>): Result<U> {
  return result.kind === "ok" ? step(result.value) : result;
}

function parseQuantity(text: string): Result<number> {
  return /^\d+$/.test(text) ? ok(Number(text)) : err(`not a number: ${text}`);
}

function checkStock(quantity: number): Result<number> {
  return quantity <= 10 ? ok(quantity) : err("only 10 in stock");
}

export function quote(text: string): Result<number> {
  return andThen(andThen(parseQuantity(text), checkStock), (q) => ok(q * 250));
}

File: tests/typesafety/python/result_chain.py
# expect: passes
# run: yes
# title: quote returns Ok with a price, or Err with the reason
# improves: nullable_chain
from collections.abc import Callable
from dataclasses import dataclass


@dataclass(frozen=True)
class Ok[T]:
    value: T


@dataclass(frozen=True)
class Err:
    reason: str


type Result[T] = Ok[T] | Err


def and_then[T, U](result: Result[T], step: Callable[[T], Result[U]]) -> Result[U]:
    match result:
        case Ok(value):
            return step(value)
        case Err():
            return result


def parse_quantity(text: str) -> Result[int]:
    return Ok(int(text)) if text.isdigit() else Err(f"not a number: {text!r}")


def check_stock(quantity: int) -> Result[int]:
    return Ok(quantity) if quantity <= 10 else Err("only 10 in stock")


def price(quantity: int) -> Result[int]:
    return Ok(quantity * 250)


def quote(text: str) -> Result[int]:
    return and_then(and_then(parse_quantity(text), check_stock), price)


assert quote("3") == Ok(750)
assert quote("99") == Err("only 10 in stock")
assert quote("many") == Err("not a number: 'many'")
